onnxruntime/core/providers/coreml/builders/impl/cast_op_builder.cc

```diff
@@ -67,14 +105,38 @@ bool CastOpBuilder::IsOpSupportedImpl(const Node& node, const OpBuilderInputPara
   return true;
 }
 
-bool CastOpBuilder::HasSupportedInputsImpl(const Node& node, const OpBuilderInputParams& /*input_params*/,
+bool CastOpBuilder::HasSupportedInputsImpl(const Node& node, [[maybe_unused]] const OpBuilderInputParams& input_params,
                                            const logging::Logger& logger) const {
   // We only check the type of input 0
   const auto& input = *node.InputDefs()[0];
+  const auto& output = *node.OutputDefs()[0];
 
-  int32_t input_type;
-  if (!GetType(input, input_type, logger))
+  int32_t input_type, output_type;
+  if (!GetType(input, input_type, logger)) {
     return false;
+  }
+  if (!GetType(output, output_type, logger)) {
+    return false;
+  }
+
+#if defined(COREML_ENABLE_MLPROGRAM)
+  if (input_params.create_mlprogram) {
+    if ((input_type == ONNX_NAMESPACE::TensorProto_DataType_INT32 ||
```

Comment:
Can't we support an input of ONNX INT64? We know it will actually be int32 inside the CoreML model.

If it's ONNX INT64 to INT32 it's a no-op (supported but use an Identity node)
If it's ONNX INT64 to fp32/fp16 I would have expected that equates to a CoreML Cast from int32 to fp32/fp16.

Reply:
Yeah, we can support it and INT64 works good.